paasta_tools/cli/cmds/get_deploygroup_history.py

```diff
@@ -0,0 +1,177 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+from typing import Collection
+from typing import Dict
+from typing import Mapping
+from typing import Tuple
+
+from humanize import naturaltime
+
+from paasta_tools.cli.cmds.mark_for_deployment import can_user_deploy_service
+from paasta_tools.cli.cmds.mark_for_deployment import get_deploy_info
+from paasta_tools.cli.utils import extract_tags
+from paasta_tools.cli.utils import figure_out_service_name
+from paasta_tools.cli.utils import lazy_choices_completer
+from paasta_tools.cli.utils import list_deploy_groups
+from paasta_tools.cli.utils import validate_given_deploy_groups
+from paasta_tools.remote_git import list_remote_refs
+from paasta_tools.utils import datetime_from_utc_to_local
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import DeploymentVersion
+from paasta_tools.utils import format_table
+from paasta_tools.utils import get_git_url
+from paasta_tools.utils import list_services
+from paasta_tools.utils import PaastaColors
+from paasta_tools.utils import parse_timestamp
+
+
+def add_subparser(subparsers: argparse._SubParsersAction) -> None:
+    list_parser = subparsers.add_parser(
+        "get-deploygroup-history",
+        help="Lists the history of a deploy group",
+        description="",  # todo
+    )
+
+    arg_service = list_parser.add_argument(
+        "-s",
+        "--service",
+        help='Name of the service to view its deploy group history (e.g. "service1")',
+        required=True,
+    )
+    arg_service.completer = lazy_choices_completer(list_services)  # type: ignore
+
+    arg_deploy_group = list_parser.add_argument(
+        "-l",
+        "--deploy-groups",
+        help="Mark one or more deploy groups to view their history (e.g. "
+        '"all.main", "all.main,all.canary").'
+        "Use the flag -a or --all-deploy-groups",
+        default="",
+        required=False,
+    )
+    arg_deploy_group.completer = lazy_choices_completer(list_deploy_groups)  # type: ignore
+
+    list_parser.add_argument(
+        "-a",
+        "--all-deploy-groups",
+        help="View history of all deploy groups for the service",
+        action="store_true",
+        required=False,
+    )
+
+    list_parser.add_argument(
+        "-y",
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+
+    list_parser.set_defaults(command=paasta_get_deploygroup_history)
+
+
+def get_versions_for_service(
+    service: str, deploy_groups: Collection[str], soa_dir: str
+) -> Mapping[DeploymentVersion, Tuple[str, str]]:
+    """Returns a dictionary of 2-tuples of the form (timestamp, deploy_group) for each version tuple of (deploy sha, image_version)"""
+    if service is None:
+        return {}
```

Comment:
is this possible? we've typed service as `service: str` so i'd imagine that mypy would say that this is unreachable code (or that there's a callsite that's passing an `Optional[str]` where `str` is expected